### DISCUSSION TITLE
Slow Elevation layer performance in 3.2 compared to 2.8

### DISCUSSION THREAD

**Discussion by tobiashermes** [pelicanmapping/osgearth#2049] [Support Q&A] [answered]
We are migrating a scene from osgEarth 2.8 to 3.2 and noticed a performance decrease with 3.2 (rex) while loading TMS elevation data. In 3.2 it always takes a moment until the data pops up. In 2.8 we had a smooth loading terrain.

For testing we created a very basic setup and there is no other data involved than the Elevation layer. The data comes from a local hard drive and was created with osgearth_conv.

Video:
https://user-images.githubusercontent.com/6120647/164713945-2dd236a7-7e5c-4326-8c58-2d9c6d1b8cbd.mp4

Earth files and debug log:
[comp_28_32_test.zip](https://github.com/gwaldron/osgearth/files/8541032/comp_28_32_test.zip)

Any ideas how to achieve a similar performance as in 2.8 with the mp-engine? 

In the 3.2 debug output there are a lot of `Replaced -340282346638528859811704183484516925440.000000 with NO_DATA_VALUE` lines. - Could this be a problem? Or maybe some configuration settings for the rex engine I am missing?

Thanks!

**Comment by tobiashermes** [comment]
If I switch off normal_maps computation I get a good performance with the rex-engine (comparable to 2.8 + mp-engine):
```
<options>
    <terrain normal_maps = "false"/>
...
``` 
Vice versa: Enabling normal maps in the mp-engine slows down loading performance (comparable to 3.3 with default rex-engine settings).

So if you migrate and experience a similar performance loss this could be one thing worth trying.

**Comment by gwaldron** [comment, marked as answer]
You can also try progressive mode:
```xml
<options>
    <terrain progressive="true"/>
```
Some people notice a perceived faster loading time with this on.

**Reply by tobiashermes** [reply]
Yes! In progressive mode it loads faster. Now the overall performance is comparable to our old scene. Thanks!

**Reply by nebsar** [reply]
How can I apply "terrain progressive = true" programmatically without an earth file?

**Reply by tobiashermes** [reply]
I haven't tried but it looks like you could use: https://github.com/gwaldron/osgearth/blob/0f589e511f6acaabc597d8d8bf237c15fe884c9a/src/osgEarth/TerrainOptions#L198

if you build your map like this:
https://github.com/gwaldron/osgearth/blob/22fe32db57f586101a495bd6de35ae69fe59d362/src/applications/osgearth_minimap/osgearth_minimap.cpp#L52

**Reply by nebsar** [reply]
Ok. I think "mapNode->getTerrainOptions().setProgressive(true);" works. Thank you.

**Reply by liu-do-program** [reply]
why is it still invalid ?
